Fix build of maintenance/0.9.x

jackvreeken opened the pull request:
Unfortunately I still have a need for Pymoca 0.9.* , and need to move to Python 3.12+ because of platform constraints that are out of my hands.

Hopefully (and definitely once #307 lands) will be able to move RTC-Tools to a more recent Pymoca version.

Commit:
Update versioneer

Remove the copy and just use pip installed version.
Versioneer calls this "build-time dependency mode" vs the "classic
vendored mode", the only option when we started using Versioneer.
Update embedded code generated by `versioneer install --no-vendor`

Versioneer, Setuptools, and/or Python 3.11 now apparently require
the use of pyproject.toml

On master we moved there, but here on the maintenance branch 0.9.x we
did not. So many settings are left in setup.cfg and setup.py .

jackvreeken committed:
codegen: Workaround MSVC NaN constantness bug

Version 10.0.26100 of the Windows SDK introduced changes to how INFINITY
and NAN are defined in `math.h`. This resulted in NAN no longer being a
compile-time constant, and codegen'd code failing compilation the most
recent `windows-latest` image. The workaround is to define
_UCRT_NOISY_NAN, which restores the old behavior.

kentrutan committed:
Remove ANTLR files from `MANIFEST.in`

Remove the `java` directory from `MANIFEST.in`. These are the files
that are included into the distribution. This directory does not exist
and produced a warning to that effect during the distribution build.

The `java` directory was previously renamed to `antlr` in the
repository and currently includes a script and an ANTLR jar file used
to generate the parser. We expect that developers wishing to change the
grammar and parser will want to do so from the repository and not the
distribution, so we remove it the ANTLR directory instead of fixing it in `MANIFEST.in`.

jackvreeken committed:
Bump minimum python to 3.9

Github Action no longer supports 3.5, and it has also been end-of-life
for over 4 years at time of writing. We bump the minimum version to
Python 3.9 (the lowest version still supported).

We also get rid of the manual version check in setup.py, which no longer
works for Python 3.10 and higher (due to "10" < "7"). We just rely on
"python_requires" from this point, which should have wide enough
adoption.

jackvreeken committed:
Parser: Update ANTLR to 4.13.1

Our ANTLR 4.7.* runtime uses `typing.io` that has been deprecated in
Python 3.12 release, so we are updating to the ANTLR version that we run
on `master` at this time.

See 65ff7987 (the commit on master) for more details. Note that that
commit does not cleanly apply, as we are missing other changes to the
grammar on this maintenance branch.

jackvreeken committed:
CI/GA: Update actions to their latest versions

See 93e113f . Also includes other changes to the `deploy` job to bring
in line with current `master`.

jackvreeken committed:
Update flake8 to latest version to avoid errors

We would get an `'EntryPoints' object has no attribute 'get'` error.
Updating flake8 to the latest version fixes it.

Only one file has new offending changes; we add it to the ignore list.

jackvreeken committed:
tests: Skip failing SymPy test

We only care about the CasADi backend, especially in this maintenance
branch.

jackvreeken committed:
setup.py/pyproject.toml: Move mostly to pyproject.toml

See (on master):
- 329bc181751747c5e78fb076e4550711ac71e99b
- 082c040a26e6b8b760d81b634fd5c1bfd493bcec
- f1eac23277545d1618b31240e08cf2cee8136643

For the antlr related changes (moving the compilation code):
- e5027d7cb6658492b045bd051f0414e025c84b54
- e34e2ad1cbf710284bbfcd027bbf825f87392679

For the SymPy pinning, see:
- fa63d48e1da1cff33383e2b14193b91abf26e962

jackvreeken commented:
Any general comments on this @kentrutan ?

kentrutan commented:
Sorry, don't have my laptop this weekend. Any general questions or can it wait till next week?

jackvreeken commented:
For sure can wait until next week, no rush :)

kentrutan commented on tools/create_conda_env.sh:
```diff
@@ -12,7 +12,7 @@ conda update -q conda
 conda info -a
 conda config --add channels conda-forge
-conda create -n pymoca python=3.5 || echo "environment already created"
+conda create -n pymoca python=3.7 || echo "environment already created"
```
I get the error below with Python 3.7. Changing to 3.9 fixes it. Seems like it should be 3.9 anyway since that is the min version with this PR.

Feel free to ignore this since this script is not included in the distribution, but I figure I would bring it up since the commit had the 3.5 -> 3.7 change.

```
PackagesNotFoundError: The following packages are not available from current channels:

  - python=3.7*

Current channels:

  - https://conda.anaconda.org/conda-forge
  - defaults
```

jackvreeken replied:
Fixed!

kentrutan commented on tools/create_conda_env.sh:
```diff
@@ -12,7 +12,7 @@ conda update -q conda
 conda config --add channels conda-forge
-conda create -n pymoca python=3.5 || echo "environment already created"
+conda create -n pymoca python=3.7 || echo "environment already created"
 conda install -n pymoca gcc jinja2 matplotlib numpy scipy sympy
```
After changing to Python 3.9, I get the error below. Removing gcc from the installed packages fixes it. Without gcc, the test suite passes and "python -m build" works fine, so I don't know what might have depended on it.

```
Channels:
 - conda-forge
 - defaults
Platform: osx-arm64
Collecting package metadata (repodata.json): done
Solving environment: failed

PackagesNotFoundError: The following packages are not available from current channels:

  - gcc

```

jackvreeken replied:
Fixed

kentrutan commented on pyproject.toml:
```diff
@@ -2,6 +2,58 @@
+# in the CasADi backend.
+casadi = ["casadi>=3.4.0", "setuptools>=60.0.0"]
+lxml = ["lxml>=3.5.0", "scipy>=0.13.3"]
+sympy = ["sympy>=0.7.6.1", "scipy>=0.13.3", "jinja2>=2.10.1"]
```
Commit ae3e226fc72a8db2e7e4a847b2c5d5a89faf08b1 message:
> There is one test in the SymPy backend that fails with these changes, so
> we skip the test as we do not care about that backend in this
> maintenance branch.

Would pinning max sympy as in the earlier commit on this branch fa63d48 also work instead of skipping the test?

jackvreeken replied:
I'll test it.

EDIT:
Yep, fixes it. I'll include in the list of commits I'm including changes of 😅

kentrutan commented on tools/create_conda_env.sh:
```diff
@@ -12,7 +12,7 @@ conda update -q conda
+conda create -n pymoca python=3.7 || echo "environment already created"
 conda install -n pymoca gcc jinja2 matplotlib numpy scipy sympy
 source activate pymoca
 # Note: casadi installed using pip since conda version currently
```
gcc is possibly needed for slycot which has to be built from source, but I only found that out after fixing the typo "eneduser" on line 24 😄. I'm now sorry for wasting time on it.

jackvreeken replied:
I'll just leave it at removing gcc. The conda script is not something I use personally (never have), and I'm not even sure it works at all anymore.

kentrutan reviewed (approved):
I checked out the .dev0 package on PyPI and it looks good as does  importing out of an editable install in a local repo. Just have a few minor comments and then should be good to go IMHO.

Probably should remove "Closes #338" in the commit message for 11ccc0f since the commit on master is actually what closed the issue.

Do you want to add antlr_build.py from 329bc181751747c5e78fb076e4550711ac71e99b and e5027d7cb6658492b045bd051f0414e025c84b54 since it was removed from setup.py? Also possibly move it into the antlr directory per e34e2ad1cbf710284bbfcd027bbf825f87392679 to be consistent with master? Also fix the distribution ala ccbe8f6da8f3462e15f8ab0fd334f00cdd6f931b if doing the others?

Also see the code comments. I don't think any of these are worth your time if you are in a hurry to get it done, so I'm going to go ahead and finish the review as approved.

jackvreeken committed:
create_conda_env: Fix running with 3.9 by removing gcc

We would get the error:

  PackagesNotFoundError: The following packages are not available from
  current channels:

Without gcc, the test suite passes and "python -m build" works fine, so
we assume everything is OK.

jackvreeken commented:
> Probably should remove "Closes https://github.com/pymoca/pymoca/issues/338" in the commit message for https://github.com/pymoca/pymoca/commit/11ccc0faf5d2d9ede6184d7a6f4e21f83e6cae57 since the commit on master is actually what closed the issue.

Done .

> antlr stuff

Also included in the `setup.py/pyproject.toml: Move mostly to pyproject.toml` commit now. Good catch!

jackvreeken commented:
Tagged an 0.9.2rc1 . I'll wait a week or two before tagging the final version. First I'll try using it by referring to it explicitly, tell others, and see if myself or they run any issues with it. I'll merge this one for now though. Those new potential issues can be their own PR.

Thanks for the review!
